meta-mbl-distro/scripts/lib/mbl/update/payload.py

```diff
@@ -0,0 +1,154 @@
+          payload.
+        * rootfs bool: True if the rootfs component should be added to the
+          payload.
+        * apps list<str|Path>: list of apps (ipk files) to add to the payload.
+
+        """
+        deploy_dir = pathlib.Path(
+            tutil.get_bitbake_conf_var("DEPLOY_DIR_IMAGE", tinfoil)
+        )
+        self.images = []
+        if bootloader_components:
+            bootloader_components_copy = bootloader_components.copy()
+            if _bootloader_one_with_kernel(bootloader_components, tinfoil):
+                if not kernel:
+                    logging.warning(
+                        "On this target the bootloader 1 component and kernel "
+                        "must be updated together. "
+                        "Adding kernel to payload..."
+                    )
+                    images.append(bootimage.BootImage(deploy_dir, tinfoil))
+
+                bootloader_components_copy.remove("1")
+
+            for bootloader_slot_number in bootloader_components_copy:
+                slot_name = "WKS_BOOTLOADER{}".format(bootloader_slot_number)
+                self.images.append(
+                    wksbootloaderslotimage.WksBootloaderSlotImage(
+                        slot_name, deploy_dir, tinfoil
+                    )
+                )
+
+        if kernel:
+            if _kernel_with_bootloader_one(bootloader_components, tinfoil):
+                if "1" not in bootloader_components:
+                    bb.warn(
+                        "On this target the bootloader 1 component and kernel "
+                        "must be updated together. "
+                        "Adding bootloader 1 component to payload..."
+                    )
+            self.images.append(bootimage.BootImage(deploy_dir, tinfoil))
+
+        if apps is not None:
+            self.images.append(appsimage.AppsImage(apps))
+
+        if rootfs is not None:
+            self.images.append(
+                rootfsimage.RootfsImage(rootfs, deploy_dir, tinfoil)
+            )
+
+    def create_payload_file(self, output_path):
+        """
+        Create an update payload.
+
+        :param output_path Path: path where we output the payload.
+        """
+        with tempfile.TemporaryDirectory() as staging_dir:
+            staging_dir_path = pathlib.Path(staging_dir)
+            swdesc_name = "sw-description"
+            swdesc.create_swdesc_file(
+                self.images, staging_dir_path / swdesc_name
+            )
+            # swupdate requires that the  sw-description file is first in the
+            # payload
+            _append_to_payload(
+                staging_dir_path, swdesc_name, output_path, create=True
+            )
+            for image in self.images:
+                image.stage(staging_dir_path)
+                _append_to_payload(
+                    staging_dir_path, image.archived_path, output_path
+                )
+
+    def create_testinfo_file(self, output_path):
```

Comment:
I suppose that the testinfo will be added to all payloads created. In the production environment will we want this information? If not, I suggest we make this optional, unless this info can be potentially used to self validate the update without any additional host tools/tests.

Reply:
There's a command line option for adding the test info.

Reply:
Thanks for clarifying.